src/System.Memory/src/System/Buffers/ArrayBufferWriter.cs

```diff
@@ -0,0 +1,177 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Buffers
+{
+    public sealed class ArrayBufferWriter<T> : IBufferWriter<T>, IDisposable
+    {
+        private T[] _rentedBuffer;
+        private int _index;
+
+        private const int MinimumBufferSize = 256;
+
+        public ArrayBufferWriter()
+        {
+            _rentedBuffer = ArrayPool<T>.Shared.Rent(MinimumBufferSize);
```

Comment:
Rent on first use vs .ctor? e.g. could be `_rentedBuffer = Array<T>.Empty();`

Reply:
What would be the benefit? With this change, we end up going through the unnecessary resize logic when _rentedBuffer == 0 on first GetSpan/GetMemory (because `sizeHint > availableSpace`)

Instead of just renting the min sized array (what we have today), you end up copying an empty buffer, clearing it, and returning it to the pool.